File: 2022/OpenCV4 tutorial/2 Get Data from Image pixel/get_data_from_image1.cs
using System;
using System.IO;
using OpenCvSharp; //导入OpenCV4
namespace get_data_from_image1
{
    class Program
    {

        static void Main(string[] args)
        {
            DirectoryInfo rootDir = Directory.GetParent(Environment.CurrentDirectory);
            string root = rootDir.Parent.Parent.FullName;
            System.IO.Directory.SetCurrentDirectory(root);
            
            // 读图
            Mat src = Cv2.ImRead("aaa.jpg", ImreadModes.Color);

            if (!src.Empty())
            {
                // 读取(317,455)的数据,相当于读455行，317列的数据
                Vec3b p = src.At<Vec3b>(455, 317);
                /*! 注意，At函数如果越界会报:“Attempted to read or write protected memory. This is often an indication that other memory is corrupt.”)*/
                Console.WriteLine( p );

                // 读取(317,455)的r通道数据
                Byte r = src.At<Vec3b>(455, 317)[2];
                Console.WriteLine( r );
            }

        }
    }
}

File: 2022/OpenCV4 tutorial/2 Get Data from Image pixel/get_data_from_image1.cpp
#include <opencv2/opencv.hpp>

using namespace cv;
int main(void)
{
    // 读图
    Mat src = imread("aaa.jpg", IMREAD_COLOR);
    
    // 读取(317,455)的数据
    Vec3b p = src.at<Vec3b>(Point2i(317,455));
    //! 注意，at函数如果越界会闪退，(程序返回1)
    std::cout<<p<<std::endl;
    // todo 另一种方式读取(317,455)的数据，相当于读455行，317列的数据
    p = src.at<Vec3b>(455,317);
    
    std::cout<<p<<std::endl;
    // 读(317,455)的R通道数据
    uchar r = src.at<Vec3b>(455,317)[2];
    std::cout<<static_cast<int>(r)<<std::endl;
    
    return 0;

}
